### PR TITLE
[sunsa] Sunsa Binding initial contribution

### PR DESCRIPTION
**Pull request by jirom**
# Description

This is a new binding to support controlling [Sunsa blinds](https://www.sunsahomes.com/).

Currently the binding requires making requests to the [Sunsa cloud APIs](https://app.swaggerhub.com/apis/Sunsa/Sunsa/1.0.4) as the local API is not available yet.

https://community.openhab.org/t/sunsa-binding/134733

# Testing
Test jar is available in:
https://openhab.jfrog.io/artifactory/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.binding.sunsa/3.3.0-SNAPSHOT/org.openhab.binding.sunsa-3.3.0-SNAPSHOT.jar

### CODE REVIEW

File: bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                final int rawPosition = getPositionAdapter(channelUID.getId())
+                        .getRawPosition(getPositionInterpolationFactor(command));
+                LOGGER.debug("Attempting update to raw position to: {}", rawPosition);
+                updatePosition(rawPosition);
+                break;
+            }
+
+            case CHANNEL_RAW_POSITION: {
+                final int rawPosition = ((DecimalType) command).intValue();
+                LOGGER.debug("Attempting update to raw position to: {}", rawPosition);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java:104]: Not sure if the logging should be here. As the framework allready logs the event. Same for other switch options

File: bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                updatePosition(rawPosition);
+                break;
+            }
+
+            case CHANNEL_BATTERY_LEVEL: {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java:109]: According to the README this is a read only channel, so i don't think it should be in this switch statement. the RefreshType command is allready handled at the top.

File: bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        if (command instanceof PercentType) {
+            rawValue = ((PercentType) command).floatValue();
+        } else if (command instanceof UpDownType) {
+            rawValue = ((UpDownType) command).as(PercentType.class).floatValue();
+        } else if (command instanceof OpenClosedType) {
+            rawValue = ((OpenClosedType) command).as(PercentType.class).floatValue();
+        } else if (command instanceof DecimalType) {
+            rawValue = ((DecimalType) command).floatValue();
+        } else {
+            throw new IllegalArgumentException("Unknown command: " + command);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java:139]: What happens if this Exception is thrown? It doesn't looks like it is handled upstream so it causses a thread crash? Similar for the other IllegalArgumentException

File: bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+        return rawValue / PercentType.HUNDRED.floatValue();
+    }
+
+    private void updatePosition(final int rawPosition) {
+        final int updatedRawPosition;
+        try {
+            updatedRawPosition = sunsaService.setDevicePosition(config.id, rawPosition);
+        } catch (final SunsaException e) {
+            LOGGER.error("Could not update position for device: {}", config.id, e);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java:150]: The logging should be removed. All info is allready present at the updateStatus. 
Should the device really be set to offline? Is there no possible way the value is rejected, while the cloud/device is perfectly healthy?

File: bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+    private boolean isConnected() {
+        return getDevice().map(Device::isConnected).orElse(false);
+    }
+
+    private Optional<Device> getDevice() {
+        try {
+            return Optional.of(sunsaService.getDevice(config.id));
+        } catch (final SunsaException e) {
+            LOGGER.error("Could not get info for device: {}", config.id, e);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.sunsa/src/main/java/org/openhab/binding/sunsa/internal/device/SunsaDeviceHandler.java:183]: Remove logging please.

File: bundles/org.openhab.binding.sunsa/.gitignore [hunk 6]
```diff
@@ -0,0 +1 @@
+/org
```
Comment (lsiepel) [inline bundles/org.openhab.binding.sunsa/.gitignore:1]: Don't think this file was meant to be part of the commit

**Review by lsiepel — CHANGES_REQUESTED** [review]
Didn't have time to look at all files, but here are some toughts.

**Review by lsiepel — CHANGES_REQUESTED** [review]
Reviewed the remaining files.

File: bundles/org.openhab.binding.sunsa/pom.xml [hunk 7]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.3.0-SNAPSHOT</version>
```
Comment (wborn) [inline bundles/org.openhab.binding.sunsa/pom.xml:10]:
```suggestion
    <version>3.4.0-SNAPSHOT</version>
```

**Comment by lsiepel** [comment]
@jirom  Are you able to proceeed with the PR? Anything you need help with?

File: bundles/org.openhab.binding.sunsa/pom.xml [hunk 8]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>4.0.0-SNAPSHOT</version>
```
Comment (wborn) [inline bundles/org.openhab.binding.sunsa/pom.xml:10]:
```suggestion
    <version>4.1.0-SNAPSHOT</version>
```

**Comment by lsiepel** [comment]
@jirom Some months have passed. are you still planning to bring this PR to the finish line?

**Comment by lsiepel** [comment]
@jirom would you be able to update this branch and remove the excess commits that should not be part of this PR?

**Comment by lsiepel** [comment]
Closed due to no response to open comments and excess commits. Feel free to re-open when this got sorted on your feature branch. Still think this binding can be a good addition to openHAB, so i hope you will manage to fix this. Ping me when you need any help to proceed.

**Pull request closed without merging**